saw-core-coq/src/Verifier/SAW/Translation/Coq/Term.hs

```diff
@@ -530,92 +579,94 @@ translateImplicitHyp ::
+      map (\(nh,nhty) -> Coq.PiImplicitBinder (Just nh) nhty) nhs
+    mkArg b = Coq.Var $ bindTransIdent b
 
--- | Find an fresh, as-yet-unused variant of the given Coq identifier.
-freshVariant :: TermTranslationMonad m => Coq.Ident -> m Coq.Ident
-freshVariant x =
-  do used <- view unavailableIdents <$> get
-     let ident0 = x
-     let findVariant i = if Set.member i used then findVariant (nextVariant i) else i
-     let ident = findVariant ident0
-     modify $ over unavailableIdents (Set.insert ident)
-     return ident
+-- | Given a list of 'LocalName's and their corresponding types (as 'Term's),
+-- return a list of explicit 'Binder's, for use representing the bound variables
+-- in 'Lambda's, 'Let's, etc.
+translateParams :: TermTranslationMonad m => [(LocalName, Term)] ->
+                   ([Coq.Binder] -> m a) -> m a
+translateParams bs m =
+  translateBinders bs (m . concat . map bindTransToBinder)
 
-nextVariant :: Coq.Ident -> Coq.Ident
-nextVariant = reverse . go . reverse
-  where
-    go :: String -> String
-    go (c : cs)
-      | c == '9'  = '0' : go cs
-      | isDigit c = succ c : cs
-    go cs = '1' : cs
+-- | Given a list of 'LocalName's and their corresponding types (as 'Term's)
+-- representing argument types and a 'Term' representing the return type,
+-- return the resulting 'Pi', with additional implicit arguments added after
+-- each instance of @isort@, @qsort@, etc.
+translatePi :: TermTranslationMonad m => [(LocalName, Term)] -> Term -> m Coq.Term
+translatePi binders body =
+  translatePiBinders binders $ \bindersT ->
+  do bodyT <- translateTermLet body
+     return $ Coq.Pi bindersT bodyT
 
+-- | Given a 'LocalName' and its type (as a 'Term'), return an explicit
+-- 'PiBinder' followed by zero or more implicit 'PiBinder's representing
+-- additonal implicit typeclass arguments, added if the given type is @isort@,
+-- @qsort@, etc.
+translatePiBinders :: TermTranslationMonad m => [(LocalName, Term)] ->
+                      ([Coq.PiBinder] -> m a) -> m a
+translatePiBinders bs m =
+  translateBinders bs (m . concat . map bindTransToPiBinder)
+
+-- | Find all subterms of a SAW core term that should be shared, and generate
+-- let-bindings in Coq to bind them to local variables. Translate SAW core term
+-- using those let-bindings for the shared subterms.
 translateTermLet :: TermTranslationMonad m => Term -> m Coq.Term
 translateTermLet t =
-  withLocalTranslationState $
-  do let counts = scTermCount False t
-     let locals = fmap fst $ IntMap.filter keep counts
-     names <- traverse (const nextName) locals
-     modify $ set sharedNames names
-     defs <- traverse translateTermUnshared locals
+  let occ_map = scTermCount False t
+      shares = IntMap.assocs $ fmap fst $ IntMap.filter keep occ_map
+      share_tms = map snd shares in
+  -- NOTE: larger terms always have later stAppIndices than their subterms, so
+  -- IntMap.elems above is guaranteed to return subterms before superterms; this
```

Comment:
I don't see a call to `IntMap.elems` above. Should this say something like "`IntMap.assocs`" instead?